docs/compact_ranges.md

```diff
@@ -0,0 +1,68 @@
+
+## Merging Compact Ranges
+
+The core property that makes compact ranges widely usable is that they are “mergeable”. Two compact ranges, `[L, M)` and `[M, R)`, can be efficiently merged into an `[L, R)` range. Consider the picture below for an intuitive understanding of how it works.
+
+![compact_ranges_merge](images/compact_ranges_merge.png)
+
+Given 2 compact ranges, `[2, 9)` and `[9, 16)`, each represented by a set of node hashes (3 green and 3 cyan nodes correspondingly), we “merge” 2 sibling nodes by computing their parent’s hash any time they are both present in the set of nodes. This process repeats until there are no siblings in the set. As a result, we get hashes of nodes `[1.1, 2.1, 3.1]` which, as it turns out, represent a compact range of `[2, 16)`.
+
+Note that, when merging 2 compact ranges, the set of “new” nodes (marked in yellow) that are generated as a side effect, forms a partial path towards the root of the tree. It can be proven that this is always the case, which is a convenient property for implementations.
+
+Merging two compact ranges can be implemented in `O(log(R-L))`, or more generally `O(log N)` time. This follows from the observation in the paragraph above, and the fact that the size of the resulting `[L, R)` compact range is limited by the same estimate.
+
+## Merkle Tree Proofs
+
+A compact range `[L, R)`, if represented by the cryptographic hashes of the corresponding nodes, can be considered a commitment to the contents of the leaves in this range. The ability to merge compact ranges is effectively the ability to merge commitments.
+
+### Inclusion Proofs Revisited
+
+A classic way of thinking about Merkle tree proofs is “vertical”, or recursive. For example, see how proofs are [defined](https://datatracker.ietf.org/doc/html/rfc6962#section-2.1) in RFC 6962. In the context of this document we rather look at Merkle trees as horizontal structures (hence *ranges*).
+
+Consider an inclusion proof for leaf `6` in the example tree below. Nodes `[3.1, 2.0, 1.2, 7]` represent a classic CT-style vertical inclusion proof.
+
+![inclusion_proof](images/inclusion_proof.png)
+
+Mind the coloring though. Nodes `[2.0, 1.2]` to the left of the path to leaf `6` are simply the compact range of `[0, 6)`. Similarly, nodes `[7, 3.1]` to the right are the compact range of `[7, 16)`.
+
+One way to verify such a proof is to run a vertical CT-style loop combining hashes, and compare the result with the expected root hash. Another way to combine the hashes and get the same result is to merge 3 compact ranges: `[0, 6)`, `[6, 7)` and `[7, 16)`. The boundary compact ranges are provided by the server as a proof, and the (trivial) middle one is constructed by the client.
```

Comment:
I don't find this section, and in particular this paragraph, very clear.  The idea of a "boundary" compact range hasn't been defined, and it might just be better to spell it out: for a leaf at index i, the server gives the ranges `[0, i)` and `[i+1, N)`, which the client can verify by merging with a middle range `[i, i+1)` formed simply as the hash of the leaf data.  This is essentially what you've said but it feels more explicit to me... maybe only to me though :).

Reply:
I rewrote this section, how does it look now?